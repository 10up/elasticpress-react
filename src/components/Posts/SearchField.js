/**
 * ElasticPress search field component
 */

import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types'; /* eslint-disable-line import/no-extraneous-dependencies */
import { replacePlaceholderInObjectValues } from '../../utils';
import { PostContext } from '../../contexts/PostContext';
import { post } from '../../api';

const SearchField = ({ placeholder, value, name, minSearchCharacters }) => {
	const [searchValue, setSearchValue] = useState(value);
	const [state, dispatch] = useContext(PostContext);

	const onChange = (event) => {
		const searchTerms = event.target.value;

		setSearchValue(searchTerms);

		dispatch({
			type: 'set_search_terms',
			payload: searchTerms,
		});

		if (searchTerms.length >= minSearchCharacters) {
			dispatch({
				type: 'set_loading',
				payload: true,
			});

			let newQuery = replacePlaceholderInObjectValues(
				state.query,
				'%SEARCH_TERMS%',
				searchTerms,
			);

			newQuery = replacePlaceholderInObjectValues(newQuery, '%PER_PAGE%', state.perPage);

			post(newQuery, state.endpoint).then((response) => {
				let newResults = [];
				let totalResults = 0;

				if (response.hits && response.hits.hits) {
					if (response.hits.total && response.hits.total.value) {
						totalResults = parseInt(response.hits.total.value, 10);
					}
					newResults = response.hits.hits.map(state.hitMap);
				}

				dispatch({
					type: 'set_results',
					payload: {
						results: newResults,
						totalResults,
						offset: state.perPage,
						searchTerms,
					},
				});

				dispatch({
					type: 'set_loading',
					payload: false,
				});
			});
		} else {
			dispatch({
				type: 'set_results',
				payload: {
					results: null,
					offset: 0,
					totalResults: null,
					searchTerms,
				},
			});
		}
	};

	return (
		<input
			type="search"
			className="search-field"
			placeholder={placeholder}
			value={searchValue}
			name={name}
			onChange={onChange}
		/>
	);
};

SearchField.defaultProps = {
	name: 's',
	placeholder: 'Search...',
	value: '',
	minSearchCharacters: 3,
};

SearchField.propTypes = {
	name: PropTypes.string,
	value: PropTypes.string,
	placeholder: PropTypes.string,
	minSearchCharacters: PropTypes.number,
};

export default SearchField;
