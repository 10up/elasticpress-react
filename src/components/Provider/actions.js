export const SET_RESULTS = 'SET_RESULTS';
export const SET_LOADING = 'SET_LOADING';
export const SET_SEARCH_TERMS = 'SET_SEARCH_TERMS';
export const REFINE = 'REFINE';
