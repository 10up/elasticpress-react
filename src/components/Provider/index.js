export * from './ElasticPressProvider';
export { default as ElasticPressProvider } from './ElasticPressProvider';
export { default as useElasticPress } from './useElasticPress';
export * from './actions';
