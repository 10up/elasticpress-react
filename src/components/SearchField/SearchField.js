/**
 * ElasticPress search field component
 */

import React, { useEffect, useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import { useSearch } from '../../hooks';

const SearchField = React.forwardRef(
	(
		{ placeholder, initialValue, name, minSearchCharacters, debounceMs, searchTerm, ...rest },
		ref,
	) => {
		const [localSearchTerm, setLocalSearchTerm] = useState('');
		const { refine } = useSearch();

		const search = useCallback(
			debounce(
				(value) => {
					refine(value, { minSearchCharacters });
				},
				debounceMs,
				{ leading: !!debounceMs, maxWait: 500 },
			),
			[refine, minSearchCharacters, debounceMs],
		);

		useEffect(() => {
			setLocalSearchTerm(searchTerm);
		}, [searchTerm]);

		// search if a initial value is provided from parent component
		useEffect(() => {
			if (initialValue) {
				search(initialValue);
			}
		}, [initialValue, search]);

		return (
			<input
				type="search"
				className="search-field"
				placeholder={placeholder}
				value={localSearchTerm || ''}
				name={name}
				onChange={(event) => {
					const { value } = event.target;
					setLocalSearchTerm(value);
					search(value);
				}}
				ref={ref}
				{...rest}
			/>
		);
	},
);

SearchField.defaultProps = {
	name: 's',
	placeholder: 'Search...',
	initialValue: '',
	minSearchCharacters: 3,
	debounceMs: 0,
	searchTerm: '',
};

SearchField.propTypes = {
	name: PropTypes.string,
	initialValue: PropTypes.string,
	placeholder: PropTypes.string,
	minSearchCharacters: PropTypes.number,
	debounceMs: PropTypes.number,
	searchTerm: PropTypes.string,
};

export default SearchField;
