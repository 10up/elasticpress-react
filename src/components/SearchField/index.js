export { default as SearchField } from './SearchField';
export { default as query } from './query';
