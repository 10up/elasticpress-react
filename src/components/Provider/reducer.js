import { SET_RESULTS, SET_LOADING, SET_SEARCH_TERMS } from './actions';
import { query as esQuery } from '../SearchField';

export const initialState = {
	results: null,
	searchTerms: null,
	perPage: 10,
	offset: 0,
	totalResults: null,
	loading: false,
	query: esQuery,
	hitMap: (hit) => {
		return hit._source;
	},
	node: null,
	indexName: null,
};

const reducer = (state, action) => {
	switch (action.type) {
		case SET_RESULTS: {
			const { append, ...payload } = action.payload;
			return {
				...state,
				...payload,
				results: append ? state.results.concat(payload.results) : payload.results,
			};
		}
		case SET_LOADING:
			return {
				...state,
				loading: action.payload,
			};
		case SET_SEARCH_TERMS:
			return {
				...state,
				searchTerms: action.payload,
			};
		default:
			return state;
	}
};

export default reducer;
