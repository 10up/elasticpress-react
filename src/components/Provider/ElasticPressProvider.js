import React, { useReducer, createContext, useCallback } from 'react';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';
import query from '../SearchField/query';

export const ElasticPressContext = createContext(initialState);

/**
 * @typedef {import('react')} React
 *
 */

/**
 * The ElasticPress Provider
 *
 * @param {object} props - The components props.
 * @param {string} [props.node] - The ElasticSearch node.
 * @param {string} [props.indexName] - The ElasticSearch index name.
 * @param {Function} [props.hitMap] - A function that maps the result of hits.
 * @param {object} [props.searchState] - The search state (optional);
 * @param {number} [props.perPage] - The number of hits per page.
 * @param {Function} [props.onSearch] - Callback function triggered whenever a function is executed
 * @param {object} [props.query] - An object describing the ElasticSearch query to perform.
 *
 * @returns {React.JSXElementConstructor}
 */
const ElasticPressProvider = ({
	children,
	node,
	indexName,
	hitMap,
	loadInitialData,
	searchState,
	resultsState,
	query,
	onSSR,
	onSearch,
}) => {
	invariant(node, 'You must specify a ElasticSearch node');
	invariant(indexName, 'You must specify a indexName');

	const [state, dispatch] = useReducer(reducer, {
		...initialState,
		search: searchState ?? initialState.state,
		results: resultsState ?? initialState.results,
	});

	/**
	 * Returns the ES endpoint for the desired query.
	 *
	 * @param {string} type The query type.
	 * @returns {string} Endpoint url
	 */
	const getEndpoint = useCallback(
		(type = 'search') => {
			if (type === 'search') {
				return `${node}/${indexName}/_doc/_search`;
			}

			return '';
		},
		[indexName, node],
	);

	const contextValue = {
		node,
		indexName,
		loadInitialData,
		state,
		hitMap,
		query,
		getEndpoint,
		dispatch,
		onSearch,
	};

	if (typeof window === 'undefined' && typeof onSSR === 'function') {
		onSSR(contextValue);
	}

	return (
		<ElasticPressContext.Provider value={contextValue}>{children}</ElasticPressContext.Provider>
	);
};

ElasticPressProvider.propTypes = {
	children: PropTypes.node.isRequired,
	searchState: PropTypes.shape({
		searchTerm: PropTypes.string,
		perPage: PropTypes.number,
	}),
	resultsState: PropTypes.shape({}),
	loadInitialData: PropTypes.bool,
	hitMap: PropTypes.func,
	query: PropTypes.object,
	node: PropTypes.string.isRequired,
	indexName: PropTypes.string.isRequired,
	onSSR: PropTypes.func,
	onSearch: PropTypes.func,
};

ElasticPressProvider.defaultProps = {
	searchState: initialState.search,
	resultsState: initialState.resultsState,
	query,
	hitMap: (hit) => {
		return hit._source;
	},
	loadInitialData: true,
	onSSR: null,
	onSearch: () => {},
};

export default ElasticPressProvider;
