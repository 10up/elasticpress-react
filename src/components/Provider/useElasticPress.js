import { useContext } from 'react';
import { ElasticPressContext } from './ElasticPressProvider';

/**
 * A simple hook to return the ElasticPressProvider state
 *
 * @returns {object}
 */
const useElasticPress = () => {
	return useContext(ElasticPressContext);
};

export default useElasticPress;
