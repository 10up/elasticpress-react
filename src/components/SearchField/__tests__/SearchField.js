import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { act } from '@testing-library/react-hooks';
import { ElasticPressProvider } from '../../Provider';
import { useSearch } from '../../../hooks';
import SearchField from '../SearchField';

jest.mock('../../../hooks');

const node = 'http://elasticpress.test/__elasticsearch';
const indexName = 'test';

// eslint-disable-next-line react/prop-types
const ElasticPressProviderWrapper = ({ children }) => {
	return (
		<ElasticPressProvider node={node} indexName={indexName}>
			{children}
		</ElasticPressProvider>
	);
};

describe('SearchField', () => {
	let refine;
	beforeEach(() => {
		refine = jest.fn();
		jest.useFakeTimers();
		useSearch.mockReturnValue({
			refine,
			search: {
				searchTerm: '',
			},
		});
	});

	it('works', () => {
		const { container } = render(<SearchField />, { wrapper: ElasticPressProviderWrapper });

		expect(container).toMatchSnapshot();
	});

	it('it calls refine if a initialValue is provided', () => {
		act(() => {
			render(<SearchField initialValue="test" />, {
				wrapper: ElasticPressProviderWrapper,
			});
		});

		jest.runAllTimers();

		expect(refine).toHaveBeenCalledWith('test', { minSearchCharacters: 3 });

		expect(refine).toHaveBeenCalledTimes(1);
	});

	it('it calls refine whenever searched value changes', () => {
		render(<SearchField name="search" />, {
			wrapper: ElasticPressProviderWrapper,
		});

		const searchField = screen.getByRole('searchbox');

		// since we're mocking useSearch we loose the controlled form input,
		// so we need to trigger one event with the full searched string
		// rather using userEvent.type
		fireEvent.change(searchField, { target: { value: 'test' } });

		jest.runAllTimers();

		expect(refine).toHaveBeenLastCalledWith('test', { minSearchCharacters: 3 });

		fireEvent.change(searchField, { target: { value: 'test test' } });
		jest.runAllTimers();

		expect(refine).toHaveBeenLastCalledWith('test test', { minSearchCharacters: 3 });
	});
});
