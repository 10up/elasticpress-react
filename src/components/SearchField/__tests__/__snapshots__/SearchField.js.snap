// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField works 1`] = `
<div>
  <input
    class="search-field"
    name="s"
    placeholder="Search..."
    type="search"
    value=""
  />
</div>
`;
